src/librbd/io/AioCompletion.cc

```diff
@@ -97,7 +95,8 @@ void AioCompletion::complete() {
     }
   }
 
-  state = AIO_STATE_CALLBACK;
+  done.store(true, std::memory_order_release);
```

Comment:
This should be done in `notify_callbacks_complete()` because a user could call `wait_for_complete()` while the callbacks are being run and we don't want to return prematurely in that case:
> There was an explicit fix to that effect, see commit [77f7c1c](https://github.com/ceph/ceph/commit/77f7c1cf829e5417eb356ebf3bf1fbc8ff7a25a5).

> * `done` would be set to true where state is set to `AIO_STATE_COMPLETE` today

Reply:
That was what I originally did, but then I noticed that some unit tests were failing because the complete state was still set to false when they called `is_complete()`, which led me to believe that the `!= AIO_STATE_PENDING` check was actually important to those tests.

I supposes I could update the state in both `complete()` and `notify_callbacks_complete()` to cover all the bases, if you don't think that would be redundant?

Reply:
> That was what I originally did, but then I noticed that some unit tests were failing because the complete state was still set to false when they called `is_complete()`, which led me to believe that the `!= AIO_STATE_PENDING` check was actually important to those tests

Do you have a list of those tests?
 
> I supposes I could update the state in both `complete()` and `notify_callbacks_complete()` to cover all the bases, if you don't think that would be redundant?

That would effectively throw out the fix made in commit 77f7c1cf829e5417eb356ebf3bf1fbc8ff7a25a5.

Reply:
I reran the make check a couple of times and it seems to fail on a different one each time, so it's not consistent and I guess there must be some sort of race condition there.

I didn't make a note of all of them sorry, but I do know that `TestLibRBD.ImagePollIO` was one of them (because I reran it locally and was unable to recreate).

Reply:
I think I see the issue -- the event socket is notified before `done`/`complete` is set.  I have opened https://github.com/ceph/ceph/pull/58591 to clean that up and separate dealing with inconsistencies in the existing code from the conversion to atomic wait/notify to gain some performance.  Let's see if tests pass there!